crypto/pkcs7/pkcs7_x509.c

```diff
@@ -428,3 +429,78 @@ PKCS7 *PKCS7_sign(X509 *sign_cert, EVP_PKEY *pkey, STACK_OF(X509) *certs,
   OPENSSL_free(der);
   return ret;
 }
+
+int PKCS7_add_certificate(PKCS7 *p7, X509 *x509)
+{
+    STACK_OF(X509) **sk;
+
+    if (p7 == NULL || x509 == NULL) {
+        OPENSSL_PUT_ERROR(PKCS7, ERR_R_PASSED_NULL_PARAMETER);
+        return 0;
+    }
+
+    switch (OBJ_obj2nid(p7->type)) {
+    case NID_pkcs7_signed:
+        sk = &(p7->d.sign->cert);
+        break;
+    case NID_pkcs7_signedAndEnveloped:
+        sk = &(p7->d.signed_and_enveloped->cert);
+        break;
+    default:
+        OPENSSL_PUT_ERROR(PKCS7, PKCS7_R_WRONG_CONTENT_TYPE);
+        return 0;
+    }
+
+    if (*sk == NULL) {
+        *sk = sk_X509_new_null();
+    }
+    if (*sk == NULL) {
+        OPENSSL_PUT_ERROR(PKCS7, ERR_R_CRYPTO_LIB);
+        return 0;
+    }
+
+    if (!sk_X509_insert(*sk, x509, 0)) {
+        X509_free(x509);
+        return 0;
+    }
+    X509_up_ref(x509);
```

Comment:
1. Minor question: Is there a reason why we insert `x509` as the first element on the stack (contrary to  `PKCS7_add_crl`)? I notice [OpenSSL 1.1.1](https://github.com/openssl/openssl/blob/OpenSSL_1_1_1-stable/crypto/pkcs7/pk7_lib.c) does as a push onto the stack for both.
2. I think we should get rid of the calls to `X509_free(x509);` here and `X509_CRL_free(crl);` on L504 below. This exists in OpenSSL because the reference count was incremented before the call.
But now that we've moved the reference call below, this could cause unintentional freeing of the user's `x509/crl` if the call fails.

Reply:
1. i'm not sure how that got in there, it looks like both [mainline](https://github.com/openssl/openssl/blob/master/crypto/pkcs7/pk7_lib.c#L298) and [1.0.2](https://github.com/openssl/openssl/blob/OpenSSL_1_0_2-stable/crypto/pkcs7/pk7_lib.c#L347) use `sk_X509_push` as well. i've updated this PR accordingly.
2. agreed.